### DISCUSSION TITLE
Windows device name and category

### DISCUSSION THREAD

**Discussion by TheStaticTurtle** [hathach/tinyusb#817] [Q&A]
Hey,

I'm not very good with USB stuff, but I'm currently developing a tool that interfaces as a generic HID device, and I was wondering if it could be possible to change the device name and/or the category in the Windows device manager.
Here it is right now:
![image](https://user-images.githubusercontent.com/17061996/117102933-9bb65d00-ad79-11eb-981c-59ddbcf4686b.png)

Is it possible for it to show under the category "BioTool" and the name "Tester"for example ?, I searched long and large, but I could find an example that do that.

Here is my current descriptor configuration and device descriptor:
```c
tusb_desc_device_t const desc_device = {
    .bLength            = sizeof(tusb_desc_device_t),
    .bDescriptorType    = TUSB_DESC_DEVICE,
    .bcdUSB             = 0x0200,
    .bDeviceClass       = TUSB_CLASS_HID,
    .bDeviceSubClass    = 0x00,
    .bDeviceProtocol    = 0x00,
    .bMaxPacketSize0    = CFG_TUD_ENDPOINT0_SIZE,

    .idVendor           = 0xCafe,
    .idProduct          = USB_PID,
    .bcdDevice          = 0x0100,

    .iManufacturer      = 0x01,
    .iProduct           = 0x02,
    .iSerialNumber      = 0x03,

    .bNumConfigurations = 0x01
};

uint8_t const desc_configuration[] = {
    // Config number, interface count, string index, total length, attribute, power in mA
    TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, TUSB_DESC_CONFIG_ATT_REMOTE_WAKEUP, 100),
                
    // Interface number, string index, protocol, report descriptor len, EP In & Out address, size & polling interval
    TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID, 0, HID_PROTOCOL_NONE, sizeof(desc_hid_report), EPNUM_HID, 0x80 | EPNUM_HID, CFG_TUD_HID_BUFSIZE, 10)
};

uint8_t const desc_hid_report[] = {
    TUD_HID_REPORT_DESC_GENERIC_INOUT(CFG_TUD_HID_BUFSIZE)
};
```

I successfully managed to change the manufacturer, product and serial with `string_desc_arr`, but I had no luck changing the device name.

I've looked into https://github.com/stacksmashing/gb-link-firmware/blob/main/usb_descriptors.c#L158, and it seems that there is some Microsoft specific configs but there isn't really documentation that I could find

**Comment by hathach** [comment]
I don't know where the Windows pull that as well, maybe you could capture the enumeration of that device. If it is windows specific thing, it is probably MS OS v2 descriptors https://docs.microsoft.com/en-us/windows-hardware/drivers/usbcon/microsoft-os-2-0-descriptors-specification .

**Reply by TheStaticTurtle** [reply]
What software do you recommand to sniff the enumeration ?
I'm pretty sure that a lot of it is windows only at least the category is, I don't think it exists for linux or mac

**Reply by hathach** [reply]
I used a hardware analyzer and don't know what software to do. I don't use windows, but I heard wireshark and/or usbmon maybe useful on linux.

**Reply by pigrew** [reply]
Thesycon USB Descriptor Dumper (https://www.thesycon.de/eng/usb_descriptordumper.shtml)
or 
USB Device Tree Viewer (https://www.uwe-sieber.de/usbtreeview_e.html)

Both of these are pretty easy to use for viewing descriptors (but these don't sniff the traffic).

**Comment by pigrew** [comment]
I believe the only way is to create a custom Windows driver for your device (INF file), and that it cannot extract a device name from the descriptors.

A Logitech mouse I have here does get a special name, but it does not match any of the names in the descriptors. The default Windows driver does come with a database of common VID/PID that it uses for names.

See:
https://www.microchip.com/forums/m476827.aspx
https://stackoverflow.com/questions/16767397/rename-a-usb-hid-device-under-windows
https://social.msdn.microsoft.com/Forums/en-US/b75746a3-fc60-4d8e-9e49-35ebb2c8c62d/usb-enumeration-display-name?forum=wdk

**Reply by TheStaticTurtle** [reply]
That would suck, but I doubt msf has such database of every possible device it just seems unreasonable, and I would even be able to submit it as it's an internal tool and I don't have a specific PID/VID
